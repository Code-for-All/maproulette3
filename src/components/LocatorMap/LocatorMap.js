import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { ZoomControl } from 'react-leaflet'
import _get from 'lodash/get'
import { latLng } from 'leaflet'
import { ChallengeLocation }
       from '../../services/Challenge/ChallengeLocation/ChallengeLocation'
import EnhancedMap from '../EnhancedMap/EnhancedMap'
import SourcedTileLayer from '../EnhancedMap/SourcedTileLayer/SourcedTileLayer'
import LayerToggle from '../EnhancedMap/LayerToggle/LayerToggle'
import WithChallengeFilters from '../HOCs/WithChallengeFilters/WithChallengeFilters'
import WithVisibleLayer from '../HOCs/WithVisibleLayer/WithVisibleLayer'
import WithMapBounds from '../HOCs/WithMapBounds/WithMapBounds'

// Setup child components with necessary HOCs
const VisibleTileLayer = WithVisibleLayer(SourcedTileLayer)

/**
 * LocatorMap presents a specially configured EnhancedMap that can be used to
 * search for challenges geographically (i.e., challenges with tasks within the
 * map bounds). It also can be used to visually represent geographic boundaries
 * while deciding on a challenge, such as when the user applies the "Near Me"
 * challenge filter.
 *
 * Initial map bounds can be provided, and the LocatorMap will also update the
 * current bounds in the redux store as the map is moved so that other components
 * (like challenge results) can apply the bounds appropriately.
 *
 * > Note: because this map both updates bounds and accepts bounds, it must be
 * > careful to avoid an infinite loop. To do this, it treats the accepted
 * > mapBounds as initial bounds and only honors new bounds if they have the
 * > fromUserAction flag set to true. This component never sets that flag
 * > itself when dispatching new map bounds.
 *
 * @see See EnhancedMap
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
export class LocatorMap extends Component {
  currentBounds = null

  shouldComponentUpdate(nextProps, nextState) {
    // We want to be careful about not constantly re-rendering, so we only
    // re-render if something meaningful changes:

    // the layer has been changed, or
    if (nextProps.layerSourceId !== this.props.layerSourceId) {
      return true
    }

    // it's the first time we've been given specific map bounds, or
    if (this.props.mapBounds.locator === null && nextProps.mapBounds.locator !== null) {
      return true
    }

    // a change in map bounds was initiated by a user action, as opposed
    // to simply navigating around in the map.
    if (_get(nextProps, 'mapBounds.locator.fromUserAction')) {
      return true
    }

    return false
  }

  /**
   * Signal a change to the current locator map bounds in response to a change
   * to the map (panning or zooming). If searching within map bounds is also
   * active, then signal that the map-bounded challenges also need to be
   * refreshed.
   *
   * @private
   */
  updateBounds = (bounds, zoom, fromUserAction=false) => {
    // If the new bounds are the same as the old, do nothing.
    if (this.currentBounds && this.currentBounds.equals(bounds)) {
      return
    }

    this.currentBounds = bounds
    this.props.setLocatorMapBounds(bounds, zoom, fromUserAction)

    if (_get(this.props, 'challengeFilter.location') ===
        ChallengeLocation.withinMapBounds) {
      this.props.updateBoundedChallenges(bounds)
    }
  }

  render() {
    return (
      <div key='locator'
           className={classNames('full-screen-map', this.props.className)}>
        <LayerToggle {...this.props} />
        <EnhancedMap center={latLng(0, 45)} zoom={3} minZoom={2} maxZoom={18}
                     setInitialBounds={false}
                     initialBounds = {_get(this.props, 'mapBounds.locator.bounds')}
                     zoomControl={false} animate={true}
                     onBoundsChange={this.updateBounds}>
          <ZoomControl position='topright' />
          <VisibleTileLayer defaultLayer={this.props.layerSourceId} />
        </EnhancedMap>
      </div>
    )
  }
}

LocatorMap.propTypes = {
  /**
   * Initial bounds at which to render the map. To avoid an infinite loop, this
   * will only be honored once, unless the fromUserAction flag is set to true
   * on updated mapBounds values.
   */
  mapBounds: PropTypes.object,
  /** Invoked when the user moves the locator map */
  setLocatorMapBounds: PropTypes.func.isRequired,
  /** layerId of default layer to display */
  layerSourceId: PropTypes.string,
  /** The currently enabled challenge filter, if any */
  challengeFilter: PropTypes.object,
}

export default WithChallengeFilters(WithMapBounds(LocatorMap))
