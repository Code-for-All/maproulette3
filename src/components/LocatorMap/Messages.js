import { defineMessages } from 'react-intl'

/**
 * Internationalized messages for use with LocatorMap
 */
export default defineMessages({
  startChallengeLabel: {
    id: "LocatorMap.controls.startChallenge.label",
    defaultMessage: "Start Challenge"
  },
})
