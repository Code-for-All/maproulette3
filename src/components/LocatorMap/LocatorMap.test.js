import React from 'react'
import { cloneDeep as _cloneDeep } from 'lodash'
import { toLatLngBounds } from '../../services/MapBounds/MapBounds'
import { LocatorMap } from './LocatorMap'

let basicProps = null

beforeEach(() => {
  basicProps = {
    mapBounds: {
      locator: {
        bounds: toLatLngBounds([0, 0, 0, 0]),
      }
    },
    layerSourceName: "foo",
    setLocatorMapBounds: jest.fn(),
  }
})

test("it renders a full screen map", () => {
  const wrapper = shallow(
    <LocatorMap {...basicProps} />
  )

  expect(wrapper.find('.full-screen-map').exists()).toBe(true)

  expect(wrapper).toMatchSnapshot()
})

test("doesn't rerender simply because the map bounds change", () => {
  const wrapper = shallow(
    <LocatorMap {...basicProps} />
  )

  const newProps = _cloneDeep(basicProps)
  newProps.mapBounds.locator.bounds = toLatLngBounds([-20, -20, 20, 20])

  expect(wrapper.instance().shouldComponentUpdate(newProps)).toBe(false)
})

test("rerenders if the map bounds change is response to an external user action", () => {
  const wrapper = shallow(
    <LocatorMap {...basicProps} />
  )

  const newProps = _cloneDeep(basicProps)
  newProps.mapBounds.locator.bounds = toLatLngBounds([-20, -20, 20, 20])
  newProps.mapBounds.locator.fromUserAction = true

  expect(wrapper.instance().shouldComponentUpdate(newProps)).toBe(true)
})

test("rerenders if the default layer name changes", () => {
  const wrapper = shallow(
    <LocatorMap {...basicProps} />
  )

  const newProps = _cloneDeep(basicProps)
  newProps.layerSourceName = 'bar'

  expect(wrapper.instance().shouldComponentUpdate(newProps)).toBe(true)
})
