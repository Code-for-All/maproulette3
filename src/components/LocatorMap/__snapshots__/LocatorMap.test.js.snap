// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a full screen map 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LocatorMap
    mapBounds={
        Object {
            "locator": Object {
              "bounds": Object {
                "_northEast": Object {
                  "lat": 0,
                  "lng": 0,
                },
                "_southWest": Object {
                  "lat": 0,
                  "lng": 0,
                },
              },
            },
          }
    }
    setLocatorMapBounds={[Function]}
    source={
        Object {
            "id": "foo",
          }
    }
    updateBoundedChallenges={[Function]}
    visibleOverlays={Array []}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "locator",
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(Connect(Connect(InjectIntl(_WithLayerSources))))
          mapBounds={
                    Object {
                              "locator": Object {
                                "bounds": Object {
                                  "_northEast": Object {
                                    "lat": 0,
                                    "lng": 0,
                                  },
                                  "_southWest": Object {
                                    "lat": 0,
                                    "lng": 0,
                                  },
                                },
                              },
                            }
          }
          setLocatorMapBounds={[Function]}
          source={
                    Object {
                              "id": "foo",
                            }
          }
          updateBoundedChallenges={[Function]}
          visibleOverlays={Array []}
/>,
        <EnhancedMap
          animate={true}
          animateFeatures={false}
          center={
                    Object {
                              "lat": 0,
                              "lng": 45,
                            }
          }
          initialBounds={
                    Object {
                              "_northEast": Object {
                                "lat": 0,
                                "lng": 0,
                              },
                              "_southWest": Object {
                                "lat": 0,
                                "lng": 0,
                              },
                            }
          }
          justFitFeatures={false}
          maxZoom={18}
          minZoom={2}
          onBoundsChange={[Function]}
          setInitialBounds={false}
          zoom={3}
          zoomControl={false}
>
          <ZoomControl
                    position="topright"
          />
          <InjectIntl(SourcedTileLayer)
                    mapBounds={
                              Object {
                                        "locator": Object {
                                          "bounds": Object {
                                            "_northEast": Object {
                                              "lat": 0,
                                              "lng": 0,
                                            },
                                            "_southWest": Object {
                                              "lat": 0,
                                              "lng": 0,
                                            },
                                          },
                                        },
                                      }
                    }
                    setLocatorMapBounds={[Function]}
                    source={
                              Object {
                                        "id": "foo",
                                      }
                    }
                    updateBoundedChallenges={[Function]}
                    visibleOverlays={Array []}
                    zIndex={1}
          />
</EnhancedMap>,
        false,
      ],
      "className": "full-screen-map",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "mapBounds": Object {
            "locator": Object {
              "bounds": Object {
                "_northEast": Object {
                  "lat": 0,
                  "lng": 0,
                },
                "_southWest": Object {
                  "lat": 0,
                  "lng": 0,
                },
              },
            },
          },
          "setLocatorMapBounds": [Function],
          "source": Object {
            "id": "foo",
          },
          "updateBoundedChallenges": [Function],
          "visibleOverlays": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "animate": true,
          "animateFeatures": false,
          "center": Object {
            "lat": 0,
            "lng": 45,
          },
          "children": Array [
            <ZoomControl
              position="topright"
/>,
            <InjectIntl(SourcedTileLayer)
              mapBounds={
                            Object {
                                          "locator": Object {
                                            "bounds": Object {
                                              "_northEast": Object {
                                                "lat": 0,
                                                "lng": 0,
                                              },
                                              "_southWest": Object {
                                                "lat": 0,
                                                "lng": 0,
                                              },
                                            },
                                          },
                                        }
              }
              setLocatorMapBounds={[Function]}
              source={
                            Object {
                                          "id": "foo",
                                        }
              }
              updateBoundedChallenges={[Function]}
              visibleOverlays={Array []}
              zIndex={1}
/>,
            Array [],
            false,
          ],
          "initialBounds": Object {
            "_northEast": Object {
              "lat": 0,
              "lng": 0,
            },
            "_southWest": Object {
              "lat": 0,
              "lng": 0,
            },
          },
          "justFitFeatures": false,
          "maxZoom": 18,
          "minZoom": 2,
          "onBoundsChange": [Function],
          "setInitialBounds": false,
          "zoom": 3,
          "zoomControl": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "position": "topright",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "mapBounds": Object {
                "locator": Object {
                  "bounds": Object {
                    "_northEast": Object {
                      "lat": 0,
                      "lng": 0,
                    },
                    "_southWest": Object {
                      "lat": 0,
                      "lng": 0,
                    },
                  },
                },
              },
              "setLocatorMapBounds": [Function],
              "source": Object {
                "id": "foo",
              },
              "updateBoundedChallenges": [Function],
              "visibleOverlays": Array [],
              "zIndex": 1,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
        ],
        "type": [Function],
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "locator",
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(Connect(Connect(InjectIntl(_WithLayerSources))))
            mapBounds={
                        Object {
                                    "locator": Object {
                                      "bounds": Object {
                                        "_northEast": Object {
                                          "lat": 0,
                                          "lng": 0,
                                        },
                                        "_southWest": Object {
                                          "lat": 0,
                                          "lng": 0,
                                        },
                                      },
                                    },
                                  }
            }
            setLocatorMapBounds={[Function]}
            source={
                        Object {
                                    "id": "foo",
                                  }
            }
            updateBoundedChallenges={[Function]}
            visibleOverlays={Array []}
/>,
          <EnhancedMap
            animate={true}
            animateFeatures={false}
            center={
                        Object {
                                    "lat": 0,
                                    "lng": 45,
                                  }
            }
            initialBounds={
                        Object {
                                    "_northEast": Object {
                                      "lat": 0,
                                      "lng": 0,
                                    },
                                    "_southWest": Object {
                                      "lat": 0,
                                      "lng": 0,
                                    },
                                  }
            }
            justFitFeatures={false}
            maxZoom={18}
            minZoom={2}
            onBoundsChange={[Function]}
            setInitialBounds={false}
            zoom={3}
            zoomControl={false}
>
            <ZoomControl
                        position="topright"
            />
            <InjectIntl(SourcedTileLayer)
                        mapBounds={
                                    Object {
                                                "locator": Object {
                                                  "bounds": Object {
                                                    "_northEast": Object {
                                                      "lat": 0,
                                                      "lng": 0,
                                                    },
                                                    "_southWest": Object {
                                                      "lat": 0,
                                                      "lng": 0,
                                                    },
                                                  },
                                                },
                                              }
                        }
                        setLocatorMapBounds={[Function]}
                        source={
                                    Object {
                                                "id": "foo",
                                              }
                        }
                        updateBoundedChallenges={[Function]}
                        visibleOverlays={Array []}
                        zIndex={1}
            />
</EnhancedMap>,
          false,
        ],
        "className": "full-screen-map",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "mapBounds": Object {
              "locator": Object {
                "bounds": Object {
                  "_northEast": Object {
                    "lat": 0,
                    "lng": 0,
                  },
                  "_southWest": Object {
                    "lat": 0,
                    "lng": 0,
                  },
                },
              },
            },
            "setLocatorMapBounds": [Function],
            "source": Object {
              "id": "foo",
            },
            "updateBoundedChallenges": [Function],
            "visibleOverlays": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "animate": true,
            "animateFeatures": false,
            "center": Object {
              "lat": 0,
              "lng": 45,
            },
            "children": Array [
              <ZoomControl
                position="topright"
/>,
              <InjectIntl(SourcedTileLayer)
                mapBounds={
                                Object {
                                                "locator": Object {
                                                  "bounds": Object {
                                                    "_northEast": Object {
                                                      "lat": 0,
                                                      "lng": 0,
                                                    },
                                                    "_southWest": Object {
                                                      "lat": 0,
                                                      "lng": 0,
                                                    },
                                                  },
                                                },
                                              }
                }
                setLocatorMapBounds={[Function]}
                source={
                                Object {
                                                "id": "foo",
                                              }
                }
                updateBoundedChallenges={[Function]}
                visibleOverlays={Array []}
                zIndex={1}
/>,
              Array [],
              false,
            ],
            "initialBounds": Object {
              "_northEast": Object {
                "lat": 0,
                "lng": 0,
              },
              "_southWest": Object {
                "lat": 0,
                "lng": 0,
              },
            },
            "justFitFeatures": false,
            "maxZoom": 18,
            "minZoom": 2,
            "onBoundsChange": [Function],
            "setInitialBounds": false,
            "zoom": 3,
            "zoomControl": false,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "position": "topright",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "mapBounds": Object {
                  "locator": Object {
                    "bounds": Object {
                      "_northEast": Object {
                        "lat": 0,
                        "lng": 0,
                      },
                      "_southWest": Object {
                        "lat": 0,
                        "lng": 0,
                      },
                    },
                  },
                },
                "setLocatorMapBounds": [Function],
                "source": Object {
                  "id": "foo",
                },
                "updateBoundedChallenges": [Function],
                "visibleOverlays": Array [],
                "zIndex": 1,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            false,
          ],
          "type": [Function],
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
